pilot/pkg/model/context.go

```diff
@@ -140,6 +143,65 @@ func (e *Environment) Version() string {
 	return ""
 }
 
+func (e *Environment) InitClusterLocalHosts() []host.Name {
+	// Create the default list of cluster-local hosts.
+	domainSuffix := e.GetDomainSuffix()
+	defaultClusterLocalHosts := make([]host.Name, 0)
+	for _, n := range defaultClusterLocalNamespaces {
+		defaultClusterLocalHosts = append(defaultClusterLocalHosts, host.Name("*."+n+".svc."+domainSuffix))
+	}
+
+	if discoveryHost, _, err := e.GetDiscoveryAddress(); err != nil {
+		log.Errorf("failed to make discoveryAddress cluster-local: %v", err)
+	} else {
+		if !strings.HasSuffix(string(discoveryHost), domainSuffix) {
+			discoveryHost += host.Name("." + domainSuffix)
+		}
+		defaultClusterLocalHosts = append(defaultClusterLocalHosts, discoveryHost)
+	}
+
+	// Collect the cluster-local hosts.
+	clusterLocalHosts := make([]host.Name, 0)
+	for _, serviceSettings := range e.Mesh().ServiceSettings {
+		if serviceSettings.Settings.ClusterLocal {
+			for _, h := range serviceSettings.Hosts {
+				clusterLocalHosts = append(clusterLocalHosts, host.Name(h))
+			}
+		} else {
+			// Remove defaults if specified to be non-cluster-local.
+			for _, h := range serviceSettings.Hosts {
+				for i, defaultClusterLocalHost := range defaultClusterLocalHosts {
+					if len(defaultClusterLocalHost) > 0 && strings.HasSuffix(h, string(defaultClusterLocalHost[1:])) {
+						// This default was explicitly overridden, so remove it.
+						defaultClusterLocalHosts[i] = ""
+					}
+				}
+			}
+		}
+	}
+
+	// Add any remaining defaults to the end of the list.
+	for _, defaultClusterLocalHost := range defaultClusterLocalHosts {
+		if len(defaultClusterLocalHost) > 0 {
+			clusterLocalHosts = append(clusterLocalHosts, defaultClusterLocalHost)
+		}
+	}
+
+	sort.Sort(host.Names(clusterLocalHosts))
+	e.clusterLocalHosts = clusterLocalHosts
+	return clusterLocalHosts
+}
+
+func (e *Environment) GetClusterLocalHosts() []host.Name {
+	return e.clusterLocalHosts
```

Comment:
Do we need to lock around this?

Reply:
It's only used during the initialization of a PushContext. So the potential race condition would be if the meshconfig was updated, and while the `InitClusterLocalHosts` was processing, a new PushContext could be created with the stale set of clusterLocalHosts.

That's definitely a race condition that also exists in the current code, but that shouldn't stop us from fixing it if necessary.

I'll defer to your understanding of how the push context is used, but my gut would say that's a small enough hole with small enough consequences that we don't need to worry about it, as I believe push contexts are meant to be relatively ephemeral anyways (and Istio follows a general consistency model) so whenever the next push is done, things will come back in line with the updated config.

Reply:
@howardjohn any thoughts around the safety of this?

Reply:
@josephpeacock perhaps we should just add a lock here to be on the safe side? My concern is that we're adding (additional?) asynchronous access since your controller is going to be accessing it as well.

Reply:
I'm definitely fine adding it. Do you think we also want it around the `InitClusterLocalHosts`?

Reply:
yeah, just lock before we actually set it.